File: 2. Structural Design Patterns/Proxy/Typescript/ProxyExample.ts
// Proxy Pattern Example in TypeScript
// Image viewer with lazy-loading proxy

interface Image { display(): void }

class RealImage implements Image {
  constructor(private filename: string){ console.log(`Loading image: ${filename}`) }
  display(): void { console.log(`Displaying: ${this.filename}`) }
}

class ProxyImage implements Image {
  private real: RealImage | null = null
  constructor(private filename: string){}
  display(): void {
    if (!this.real) this.real = new RealImage(this.filename)
    this.real.display()
  }
}

;(function main(){ const img: Image = new ProxyImage('photo.jpg'); img.display(); img.display() })()

export {}

File: 2. Structural Design Patterns/Proxy/Kotlin/ProxyExample.kt
// Proxy Pattern Example in Kotlin
// Image viewer with lazy-loading proxy

interface Image { fun display() }

class RealImage(private val filename: String) : Image {
    init { println("Loading image: $filename") }
    override fun display() = println("Displaying: $filename")
}

class ProxyImage(private val filename: String) : Image {
    private var real: RealImage? = null
    override fun display() {
        if (real == null) real = RealImage(filename)
        real!!.display()
    }
}

fun main(){ val img: Image = ProxyImage("photo.jpg"); img.display(); img.display() }
